cmd/kpod/ps.go

```diff
@@ -0,0 +1,350 @@
+		ArgsUsage:   "",
+	}
+)
+
+type outStruct struct {
+	output []psOutputParams
+	opts   psOptions
+}
+
+func (sOut outStruct) Out() error {
+	var w *tabwriter.Writer
+	if len(sOut.output) > 0 {
+		w = tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
+		defer w.Flush()
+
+		if !sOut.opts.quiet {
+			if sOut.opts.size {
+				fmt.Fprintln(w, "CONTAINER ID\t\tIMAGE\t\tCOMMAND\t\tCREATED\t\tSTATUS\t\tPORTS\t\tNAMES\t\tSIZE\t")
+			} else {
+				fmt.Fprintln(w, "CONTAINER ID\t\tIMAGE\t\tCOMMAND\t\tCREATED\t\tSTATUS\t\tPORTS\t\tNAMES\t")
+			}
+		}
+
+		for _, ctr := range sOut.output {
+			createdTime := units.HumanDuration(time.Since(ctr.CreatedAt)) + " ago"
+			if sOut.opts.quiet {
+				fmt.Fprintln(w, ctr.ID+"\t\t")
+			} else if sOut.opts.size {
+				fmt.Fprintln(w, ctr.ID+"\t\t"+ctr.Image+"\t\t"+ctr.Command+"\t\t"+createdTime+"\t\t"+ctr.Status+"\t\t"+ctr.Ports+"\t\t"+ctr.Names+"\t\t"+ctr.Size+"\t")
+			} else {
+				fmt.Fprintln(w, ctr.ID+"\t\t"+ctr.Image+"\t\t"+ctr.Command+"\t\t"+createdTime+"\t\t"+ctr.Status+"\t\t"+ctr.Ports+"\t\t"+ctr.Names+"\t")
+			}
+		}
+	}
+	return nil
+}
+
+func toGenericPS(params []psOutputParams) []interface{} {
+	genericParams := make([]interface{}, len(params))
+	for i, v := range params {
+		genericParams[i] = interface{}(v)
+	}
+	return genericParams
+}
+
+func psCmd(c *cli.Context) error {
+	config, err := getConfig(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get config")
+	}
+	server, err := libkpod.New(config)
+	if err != nil {
+		return errors.Wrapf(err, "error creating server")
+	}
+	if err := server.Update(); err != nil {
+		return errors.Wrapf(err, "error updating list of containers")
+	}
+
+	if len(c.Args()) > 0 {
+		return errors.Errorf("too many argument, ps takes no arguments")
+	}
+
+	opts := psOptions{
+		all:     c.Bool("all"),
+		filter:  c.String("filter"),
+		format:  c.String("format"),
+		last:    c.Int("last"),
+		latest:  c.Bool("latest"),
+		noTrunc: c.Bool("no-trunc"),
+		quiet:   c.Bool("quiet"),
+		size:    c.Bool("size"),
+	}
+
+	var params *libkpod.FilterParamsPS
+	if opts.filter != "" {
+		params, err = libkpod.ParseFilter(opts.filter, server)
+		if err != nil {
+			return errors.Wrapf(err, "error parsing filter")
+		}
+	} else {
+		params = nil
+	}
+
+	containerList, err := libkpod.GetContainersMatchingFilter(server, params)
+	if err != nil {
+		return errors.Wrapf(err, "could not get list of containers matching filter")
+	}
+
+	return psOutput(containerList, server, opts)
+}
+
+func psOutput(containers []*libkpod.ContainerData, server *libkpod.ContainerServer, opts psOptions) error {
+	var (
+		output           []psOutputParams
+		containersOutput []*libkpod.ContainerData
+		status           string
+		ctrID            string
+		command          string
+		runningFor       string
+		imageName        string
+		mounts           string
+		ports            string
+		size             string
+		labels           string
+	)
+
+	if len(containers) > 0 {
+		if opts.latest {
```

Comment:
You need to check if opts.latest, opts.last or opts.all are specified then throw an error before getting here. then you can simpify the code without all of these else.if lines.

They are mutually exclusive commands so if a user specified them together then error out.